### PR TITLE
[Merged by Bors] - feat: compute `Monoid.exponent` as the lcm over pi types and `Prod`

### PR DESCRIPTION
**Pull request by j-loreaux**
---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/GroupTheory/Exponent.lean [hunk 1]
```diff
@@ -386,3 +387,51 @@ theorem card_dvd_exponent_pow_rank' {n : ℕ} (hG : ∀ g : G, g ^ n = 1) :
...
 end CommGroup
+
+section PiProd
+
+open Finset Monoid
+
+@[to_additive]
+theorem Monoid.exponent_pi_zero {ι : Type*} {M : ι → Type*} [∀ i, Monoid (M i)] {j : ι}
+    (hj : exponent (M j) = 0) : exponent ((i : ι) → M i) = 0 := by
+  rw [@exponent_eq_zero_iff, ExponentExists] at hj ⊢
```
Comment (eric-wieser) [inline Mathlib/GroupTheory/Exponent.lean:398]: Is this true as an `Iff` with an existential?

File: Mathlib/GroupTheory/Exponent.lean [hunk 2]
```diff
@@ -386,3 +387,51 @@ theorem card_dvd_exponent_pow_rank' {n : ℕ} (hG : ∀ g : G, g ^ n = 1) :
...
+    ext i
+    rw [Pi.pow_apply, Pi.one_apply, ← orderOf_dvd_iff_pow_eq_one]
+    apply dvd_trans (Monoid.order_dvd_exponent (m i))
+    exact Finset.dvd_lcm (mem_univ i)
+  · apply Finset.lcm_dvd fun i _ ↦ ?_
+    exact MonoidHom.exponent_dvd (f := Pi.evalMonoidHom (M ·) i) (Function.surjective_eval i)
+
+@[to_additive]
+theorem Monoid.exponent_prod {M₁ M₂ : Type*} [Monoid M₁] [Monoid M₂] :
+    exponent (M₁ × M₂) = Nat.lcm (exponent M₁) (exponent M₂) := by
```
Comment (eric-wieser) [inline Mathlib/GroupTheory/Exponent.lean:426]: Is this (and the pi version) worthy of the module docstring?

File: Mathlib/GroupTheory/Exponent.lean (hunk 1 above)
Comment (eric-wieser) [reply Mathlib/GroupTheory/Exponent.lean:398]: (and is the `Prod` version too, with an `or`?)

Comment (j-loreaux) [reply Mathlib/GroupTheory/Exponent.lean:398]: Note: the version I provided here does *not* assume `Fintype ι` as `Monoid.exponent_pi` does. Without that, the `↔` does not hold (consider the product `Π n, ZMod n` for instance).

With the `Fintype ι` condition, then yes, it's true, but it follows immediately from `Finset.lcm_eq_zero_iff`, and I would expect users to use that if they want this result. Similarly, the prod version holds, but follows immediately from `lcm_eq_zero_iff` (once I switch from `Nat.lcm` to `lcm`, which I just realized I should have done).

So, these are the lemmas, but I propose *not* including them:
```lean
@[to_additive]
lemma Monoid.exponent_pi_zero_iff {ι : Type*} [Fintype ι] {M : ι → Type*} [∀ i, Monoid (M i)] :
    exponent ((i : ι) → M i) = 0 ↔ ∃ i, exponent (M i) = 0 := by
  rw [Monoid.exponent_pi, Finset.lcm_eq_zero_iff]
  simp

@[to_additive]
theorem Monoid.exponent_prod_zero_iff {M₁ M₂ : Type*} [Monoid M₁] [Monoid M₂] :
    exponent (M₁ × M₂) = 0 ↔ (exponent M₁ = 0) ∨ (exponent M₂ = 0) := by
  rw [exponent_prod, lcm_eq_zero_iff]
```

File: Mathlib/GroupTheory/Exponent.lean (hunk 2 above)
Comment (j-loreaux) [reply Mathlib/GroupTheory/Exponent.lean]: done, and I added docstrings too.

File: Mathlib/GroupTheory/Exponent.lean [hunk 3]
```diff
@@ -386,3 +392,58 @@ theorem card_dvd_exponent_pow_rank' {n : ℕ} (hG : ∀ g : G, g ^ n = 1) :
 #align card_dvd_exponent_nsmul_rank' card_dvd_exponent_nsmul_rank'
 
 end CommGroup
+
+section PiProd
+
+open Finset Monoid
+
+@[to_additive]
+theorem Monoid.exponent_pi_zero {ι : Type*} {M : ι → Type*} [∀ i, Monoid (M i)] {j : ι}
```
Comment (eric-wieser) [inline Mathlib/GroupTheory/Exponent.lean:401]:
```suggestion
theorem Monoid.exponent_pi_eq_zero {ι : Type*} {M : ι → Type*} [∀ i, Monoid (M i)] {j : ι}
```

**Comment by eric-wieser** [comment]
bors d+

**Comment by j-loreaux** [comment]
bors merge

**Pull request closed without merging**